text/0000-multi-api-lifecycle-descriptor.md

```diff
@@ -0,0 +1,172 @@
+
+The lifecycle will set `io.buildpacks.lifecycle.version` and `io.buildpacks.lifecycle.apis` on all future released lifecycle images.
+
+Platforms may use the contents of the `io.buildpacks.lifecycle.apis` label when deciding which `CNB_PLATFORM_API` to use, or validating whether a buildpack is compatible with a given builder image.
+
+### `pack`
+`pack inspect-builder` should display deprecated, supported, and experimental API information when it is available.
+
+## Backwards Compatibility
+
+Currently lifecycle API and verison information is contained in the `io.buildpacks.builder.metadata` label. The following is an abridged example label
+```
+{
+  "buildpacks": [],
+  "created-by": {}
+  "description": "",
+  "lifecycle": {
+    "version": "<lifecycle version>",
+    "api": {
+      "buildpack": "<buildpack API version>",
+      "platform": "<platform API version>"
+    }
+  },
+  "stack": {},
+}
+```
+For backwards compatibility `pack create-builder` should continue to populate the above label, setting `lifecycle.api.buildpack`, and `lifecycle.api.platform` to the corresponding supported `min`.
+
+This will allow platforms the read the metadata in it's existing format to update gracefully. Eventually this compatibility layer will be removed.
+
+The lifecycle descriptor file itself will not preserve backwards compatibility. Most platforms consume builders rather than directly consuming lifecycle artifacts, and therefore the builder label is the important compatibility concern. Those platforms that do consume lifecycle artifacts control which versions of the lifecycle they bring in and can update their behavior to parse the new format for newer lifecycles.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Understanding buildpack/platform API support will require more documentation and education.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- We could have `deprected` and `supported` lists instead of using `min` and `max` to define ranges.
+- We could stick with a single `api.buildpack` and `api.platform` in the lifecycle descriptor and let platforms infer support ranges based on the guidelines outlined in [RFC-0041](https://github.com/buildpacks/rfcs/blob/main/text/0041-api-version-compat.md#non-breaking-0x-api-versions)
+- Instead of introducing deprecated APIs we could commit to never deprecating or removing support for older APIs
+
+# Prior Art
+[prior-art]: #prior-art
+
+* Experimental APIs are similar to [alpha level APIs](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning) in Kubernetes.
+* https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Should the lifecycle warn when an experimental API is used?
+- Should we support a range of experimental APIs instead of a single API?
```

Comment:
I'm not sure we'd have multiple versions. Is experimental a version or a mode?

Reply:
versions are modes right? Not sure I follow the question.

Reply:
I just mean a generic `experiment` version vs an explicit number.

Reply:
I like an explicit number better so that when a new lifecycle lands and the experiment becomes supported you don't need to change the value of `CNB_PLATFORM_API` to avoid weird behavior (i.e. when there is no experiment or it's a totally new API).

Reply:
is that worse b/c if that new number isn't an experiment we don't promise that it will behave the same?

Reply:
I think it is better. If we are promoting an experimental API to a supported API the diff between the behavior of the final iteration of experimental version and the supported version is likely to small (certainly smaller than the diff between that and the next experimental version, if there even is one).

Reply:
But based on our guarantees we're still implying there could be a difference and that's just the cost of using the experimental API even from `1.0-alpha1` vs `1.0-alpha2`. My thoughts that if someone doesn't even know that upgrading where they're now suddenly on a stable API could be pretty surprising.

Reply:
>  the diff between the behavior of the final iteration of experimental version and the supported version is likely to small

I don't want to hold ourselves to this. We should be able to make dramatic changes in behavior b/w experiment and official feature

Reply:
Following up on this from the WG meeting today. We settled on allowing both "less" experimental APIs like platform `0.4` where the feature set is much more well known as well as a `1.0` but can designate with alphanumeric characters like "1.0-alpha1". This means extending the API versioning beyond `X.Y` to `X.Y-<alphanumeric>` but only for experimental versions.

Reply:
Resolving because experimental has been removed entirely